### PR TITLE
Ticket validation certificates selection screen (EXPOSUREAPP-10483)

### PR DESCRIPTION
**Pull request by SamuraiKek**
UI for certificates selection.

Binding and navigation to consent2 should in theory be covered too.

### CODE REVIEW

File: Corona-Warn-App/src/main/res/values-de/dcc_ticketing_strings.xml [hunk 1]
```diff
@@ -56,4 +56,27 @@
     <!-- XTXT: not in AllowList dialog message -->
     <string name="dcc_ticketing_not_in_allow_list_dialog_message">Der Anbieter ist uns nicht bekannt. Bitte überprüfen Sie, ob der QR-Code wirklich von dem Anbieter stammt und ob Sie ihm Ihr Zertifikat zur Verfügung stellen wollen.</string>
```
Comment (KathrinLuecke) [inline Corona-Warn-App/src/main/res/values-de/dcc_ticketing_strings.xml:57]: Not part of this PR, but this sentence should be updated. It was changed since it has been implemented. Updated version: "Der Anbieter ist uns nicht bekannt. Aus Sicherheitsgründen ist die Überprüfung bei Ticketbuchung nur für bekannte Anbieter möglich. Bitte kontaktieren Sie den Anbieter, der ihnen dieses QR-Code bereitgstellt hat."

**Review by KathrinLuecke — DISMISSED** [review]
UA approved

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/dccticketing/ui/certificateselection/DccTicketingCertificateSelectionFragment.kt [hunk 2]
```diff
@@ -1,24 +1,120 @@
...
+
+    @Inject lateinit var viewModelFactory: CWAViewModelFactoryProvider.Factory
 
     private val args by navArgs<DccTicketingCertificateSelectionFragmentArgs>()
+    private val binding: FragmentDccTicketingCertificateSelectionBinding by viewBinding()
+    private val viewModel: DccTicketingCertificateSelectionViewModel by cwaViewModelsAssisted(
+        factoryProducer = { viewModelFactory },
+        constructorCall = { factory, _ ->
+            factory as DccTicketingCertificateSelectionViewModel.Factory
+            factory.create(args.transactionContext)
```
Comment (mtwalli) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/dccticketing/ui/certificateselection/DccTicketingCertificateSelectionFragment.kt:34]: I know this was intended from the beginning to use `navArgs`, but let us use `DccTicketingSharedViewModel` as in `ConsentI`. this viewModel shares the updated `TransactionContext` in dcc_tickeing_flow.

Why?
TransactionContext data might have size of 10 MB or more (atm it can get so big, near future we have to set a limit up to 10 MB)

this amount of data in Bundle is way to big and exceeds the limit of how much data can passed using bundle see 
https://developer.android.com/guide/components/activities/parcelables-and-bundles#sdba
>When sending data via an intent, you should be careful to limit the data size to a few KB. Sending too much data can cause the system to throw a TransactionTooLargeException exception.

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/dccticketing/ui/certificateselection/DccTicketingCertificateSelectionFragment.kt [hunk 3]
```diff
@@ -1,24 +1,120 @@
...
 
-    override fun onCreateView(
-        inflater: LayoutInflater,
-        container: ViewGroup?,
-        savedInstanceState: Bundle?
-    ): View? {
-        // Inflate the layout for this fragment
-        // TODO Pass args.transactionContext.accessTokenPayload to the view model
-        return inflater.inflate(R.layout.fragment_dcc_ticketing_certificate_selection, container, false)
+    private fun returnToScreenWhereScannerWasOpened() {
```
Comment (mtwalli) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/dccticketing/ui/certificateselection/DccTicketingCertificateSelectionFragment.kt:109]: Do we need to do this ? 

I think when you navigate from `consentI` to selection screen and pop to `consentI`

1- `ConsentI ------- popuo to consentI -> selection screen` 
2- Close in selection screen just use `popBackStack ` it will work and land where use started

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/dccticketing/ui/certificateselection/DccTicketingCertificateSelectionFragment.kt [hunk 4]
```diff
@@ -1,24 +1,120 @@
...
+
+        binding.apply {
+            toolbar.setNavigationOnClickListener { returnToScreenWhereScannerWasOpened() }
+            recyclerView.apply {
+                adapter = certificatesAdapter
+                addItemDecoration(TopBottomPaddingDecorator(topPadding = R.dimen.spacing_tiny))
+            }
+            viewModel.events.observe(viewLifecycleOwner) { event ->
+                when (event) {
+                    is NavigateToConsentTwoFragment -> doNavigate(
```
Comment (mtwalli) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/dccticketing/ui/certificateselection/DccTicketingCertificateSelectionFragment.kt:50]: Please handle close event here too ? I will be easy to test it

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/dccticketing/ui/certificateselection/DccTicketingCertificateSelectionFragment.kt [hunk 5]
```diff
@@ -1,24 +1,120 @@
...
+                                event.selectedCertificateContainerId
+                            )
+                    )
+                }
+            }
+        }
+
+        lifecycleScope.launchWhenStarted {
+            val certificatesList = viewModel.getCertificates()
+            certificatesAdapter.update(certificatesList)
```
Comment (mtwalli) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/dccticketing/ui/certificateselection/DccTicketingCertificateSelectionFragment.kt:63]: Can we use live data instead and do the logic in the viewmodel? we pass the TransactionContext to the viewmodel as argument and you can in the  ViewModel  filter the certifcates and publish them to the fragment using some live data

```kotlin 
 class DccTicketingCertificateSelectionViewModel @AssistedInject constructor(
    .....
) : CWAViewModel(dispatcherProvider) {
    val certificats = liveData<Set<CwaCovidCertificate>> {
        emit(dccTicketingCertificateFilter.filter(transactionContext.accessTokenPayload?.vc))
    }
```

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/dccticketing/ui/certificateselection/DccTicketingCertificateSelectionHelperFunctions.kt [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+            FilterType.TEST.type,
+            FilterType.RA_TEST.type -> context.getString(R.string.test_certificate_name)
+            else -> ""
+        }
+    }
+
+fun getFullName(familyName: String?, givenName: String?): String =
+    listOf(familyName, givenName).joinToString("<<")
+
+private enum class FilterType(val type: String) {
```
Comment (mtwalli) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/dccticketing/ui/certificateselection/DccTicketingCertificateSelectionHelperFunctions.kt:21]: We have one already in case you want to use please use the same one , maybe you move into an enum class and can be used in both cases ?

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/dccticketing/ui/certificateselection/DccTicketingCertificateSelectionFragment.kt [hunk 7]
```diff
@@ -1,24 +1,120 @@
...
+                    } ?: ""
+                    birthDate.text = getString(R.string.dcc_ticketing_certificate_birthday)
+                        .format(args.transactionContext.accessTokenPayload?.vc?.dob)
+                    standardizedName.text = getFullName(
+                        args.transactionContext.accessTokenPayload?.vc?.fnt,
+                        args.transactionContext.accessTokenPayload?.vc?.gnt
+                    )
+                    requiredCertificatesTitle.text =
+                        getString(R.string.dcc_ticketing_certificates_meeting_requirements_title)
+                }
```
Comment (mtwalli) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/dccticketing/ui/certificateselection/DccTicketingCertificateSelectionFragment.kt:104]: can we improve it if constraint layout groups is used ?
noMatchingCertificatesGroup.isVisible = certificates.isEmpty 

matchingCertificatesGroup.isVisible = certificates.isNotEmpty 

I see many views set visibility based on that. and maybe small functions to make readable ?

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/dccticketing/ui/certificateselection/DccTicketingCertificateSelectionFragment.kt [hunk 8]
```diff
@@ -1,24 +1,120 @@
...
+import de.rki.coronawarnapp.util.ui.popBackStack
+import de.rki.coronawarnapp.util.ui.viewBinding
+import de.rki.coronawarnapp.util.viewmodel.CWAViewModelFactoryProvider
+import de.rki.coronawarnapp.util.viewmodel.cwaViewModelsAssisted
+import setTextWithUrl
+import java.util.Locale
+import javax.inject.Inject
 
-class DccTicketingCertificateSelectionFragment : Fragment() {
+class DccTicketingCertificateSelectionFragment :
```
Comment (mtwalli) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/dccticketing/ui/certificateselection/DccTicketingCertificateSelectionFragment.kt:23]: I don't see any screenshots test :D

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/dccticketing/ui/certificateselection/DccTicketingCertificateSelectionHelperFunctions.kt [hunk 9]
```diff
@@ -0,0 +1,27 @@
+package de.rki.coronawarnapp.dccticketing.ui.certificateselection
+
+import android.content.Context
+import de.rki.coronawarnapp.R
+
+fun getRequestedCertificateTypes(requestedCertificatesList: List<String>, context: Context, separator: String): String =
```
Comment (BMItr) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/dccticketing/ui/certificateselection/DccTicketingCertificateSelectionHelperFunctions.kt:6]: You could use LazyString and CachedString, so you dont have to pass a context

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/dccticketing/ui/certificateselection/DccTicketingCertificateSelectionViewModel.kt [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+class DccTicketingCertificateSelectionViewModel @AssistedInject constructor(
+    dispatcherProvider: DispatcherProvider,
+    private val dccTicketingCertificateFilter: DccTicketingCertificateFilter,
+    @Assisted private val transactionContext: DccTicketingTransactionContext
+) : CWAViewModel(dispatcherProvider) {
+
+    val events = SingleLiveEvent<DccTicketingCertificateSelectionEvents>()
+    private val validationCondition = transactionContext.accessTokenPayload?.vc
+
+    suspend fun getCertificates() = dccTicketingCertificateFilter.filter(validationCondition).map { certificate ->
```
Comment (BMItr) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/dccticketing/ui/certificateselection/DccTicketingCertificateSelectionViewModel.kt:27]: Please use a flow and livedata

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/dccticketing/ui/certificateselection/DccTicketingTestCard.kt [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+        val certificate = curItem.certificate
+        root.setOnClickListener { curItem.onClick }
+        certificateDate.text = context.getString(
+            R.string.test_certificate_sampled_on,
+            certificate.sampleCollectedAt.toUserTimeZone().toShortDayFormat()
+        )
+
+        when (certificate.rawCertificate.test.testType) {
+            // PCR Test
+            "LP6464-4" -> R.string.test_certificate_pcr_test_type
```
Comment (BMItr) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/dccticketing/ui/certificateselection/DccTicketingTestCard.kt:33]: Please create a const for that

File: Corona-Warn-App/src/main/res/layout/dcc_ticketing_recovery_card.xml [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+        android:hyphenationFrequency="normal"
+        android:text="@string/recovery_certificate_name"
+        android:textSize="18sp"
+        android:textStyle="bold"
+        app:layout_constraintEnd_toEndOf="parent"
+        app:layout_constraintEnd_toStartOf="@id/arrow"
+        app:layout_constraintStart_toEndOf="@id/certificate_bg"
+        app:layout_constraintTop_toTopOf="@id/certificate_bg" />
+
+    <ImageView
```
Comment (BMItr) [inline Corona-Warn-App/src/main/res/layout/dcc_ticketing_recovery_card.xml:50]: Pls be constant with app:srcCompat

**Comment by BMItr** [comment]
card layouts look pretty equal, so you could reuse them. (optional)

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/dccticketing/ui/certificateselection/DccTicketingCertificateSelectionHelperFunctions.kt (hunk 6 above)
Comment (SamuraiKek) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/dccticketing/ui/certificateselection/DccTicketingCertificateSelectionHelperFunctions.kt:21]: Yep, fair enough will make an enum class for both cases.

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/dccticketing/ui/certificateselection/DccTicketingCertificateSelectionFragment.kt (hunk 8 above)
Comment (SamuraiKek) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/dccticketing/ui/certificateselection/DccTicketingCertificateSelectionFragment.kt:23]: Left them for the separate subtask :D

**Comment by SamuraiKek** [comment]
> card layouts look pretty equal, so you could reuse them. (optional)

Yes, I've had a discussion with @mtwalli about this and we decided it was best to have separate ones for this selection screen.

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/dccticketing/ui/certificateselection/DccTicketingCertificateSelectionFragment.kt (hunk 7 above)
Comment (SamuraiKek) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/dccticketing/ui/certificateselection/DccTicketingCertificateSelectionFragment.kt:104]: Added the groups in the xml's but haven't gotten the change to use them here.

File: Corona-Warn-App/src/androidTest/java/de/rki/coronawarnapp/dccticketing/ui/certificateselection/DccTicketingCertificateSelectionFragmentTest.kt [hunk 13]
```diff
@@ -0,0 +1,128 @@
...
+            listOf(
+                DccTicketingNoValidCertificateHeaderCard.Item(),
+                DccTicketingNoValidCertificateCard.Item(vc),
+                DccTicketingNoValidCertificateFaqCard.Item()
+            )
+
+        )
+        launchFragmentInContainer2<DccTicketingCertificateSelectionFragment>(
+            testNavHostController = navController
+        )
```
Comment (mtwalli) [inline Corona-Warn-App/src/androidTest/java/de/rki/coronawarnapp/dccticketing/ui/certificateselection/DccTicketingCertificateSelectionFragmentTest.kt:96]: @AlexanderAlferov  you can pass `testNavController`   as https://github.com/corona-warn-app/cwa-app-android/pull/4439/files#diff-e4121ff9898e0ed0575c9dd9ae6e547812252047c5251864ca97eceadcb52cc9R61-R69
with the current destination ofc

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/dccticketing/ui/certificateselection/DccTicketingCertificateSelectionHelperFunctions.kt [hunk 14]
```diff
@@ -0,0 +1,28 @@
...
+import android.content.Context
+import de.rki.coronawarnapp.R
+import de.rki.coronawarnapp.dccticketing.core.certificateselection.DccTicketingCertificatesFilterType
+import de.rki.coronawarnapp.dccticketing.core.certificateselection.DccTicketingCertificatesFilterType.PCR_TEST
+import de.rki.coronawarnapp.dccticketing.core.certificateselection.DccTicketingCertificatesFilterType.RA_TEST
+import de.rki.coronawarnapp.dccticketing.core.certificateselection.DccTicketingCertificatesFilterType.RECOVERY
+import de.rki.coronawarnapp.dccticketing.core.certificateselection.DccTicketingCertificatesFilterType.TEST
+import de.rki.coronawarnapp.dccticketing.core.certificateselection.DccTicketingCertificatesFilterType.VACCINATION
+
+fun Context.certificateTypesText(certificateTypes: List<String>, separator: String = ", "): String =
```
Comment (BMItr) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/dccticketing/ui/certificateselection/DccTicketingCertificateSelectionHelperFunctions.kt:12]: Would be better to use LazyString, so the context is only needed when accessing the String.

File: Corona-Warn-App/src/main/res/layout/dcc_ticketing_no_valid_certificate_header_card.xml [hunk 15]
```diff
@@ -0,0 +1,41 @@
+<androidx.constraintlayout.widget.ConstraintLayout xmlns:android="http://schemas.android.com/apk/res/android"
+    xmlns:app="http://schemas.android.com/apk/res-auto"
+    android:layout_width="match_parent"
+    android:layout_height="wrap_content">
+    <TextView
+        android:id="@+id/selection_title"
+        style="@style/materialSubtitleSixteen"
+        android:layout_width="0dp"
+        android:layout_height="wrap_content"
+        android:layout_marginStart="24dp"
```
Comment (BMItr) [inline Corona-Warn-App/src/main/res/layout/dcc_ticketing_no_valid_certificate_header_card.xml:10]: Use marginHorizontal (optional)

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/dccticketing/ui/certificateselection/DccTicketingCertificateSelectionHelperFunctions.kt (hunk 14 above)
Comment (mtwalli) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/dccticketing/ui/certificateselection/DccTicketingCertificateSelectionHelperFunctions.kt:12]: this is accessed from the view holder, so it when actually used

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/dccticketing/core/certificateselection/DccTicketingCertificateFilter.kt [hunk 16]
```diff
@@ -41,12 +46,12 @@ class DccTicketingCertificateFilter @Inject constructor(
...
-                FilterType.VACCINATION -> vaccinationCerts
-                FilterType.RECOVERY -> recoveryCerts
-                FilterType.TEST -> testCerts
-                FilterType.PCR_TEST -> testCerts.filter { it.rawCertificate.test.testType == "LP6464-4" }
-                FilterType.RA_TEST -> testCerts.filter { it.rawCertificate.test.testType == "LP217198-3" }
+            when (DccTicketingCertificatesFilterType.typeOf(type)) {
+                VACCINATION -> vaccinationCerts
+                RECOVERY -> recoveryCerts
+                TEST -> testCerts
+                PCR_TEST -> testCerts.filter { it.rawCertificate.test.testType == "LP6464-4" }
```
Comment (BMItr) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/dccticketing/core/certificateselection/DccTicketingCertificateFilter.kt:53]: We should create Constants for that and reuse them
(as done in the card)

**Review by BMItr — DISMISSED** [review]
code lgtm now

File: Corona-Warn-App/src/main/res/values-de/dcc_ticketing_strings.xml [hunk 17]
```diff
@@ -56,4 +56,31 @@
...
+    <!--XTXT: Certificate Selection Screen: FAQ Link EN -->
+    <string name="dcc_ticketing_faq_link_english">"https://www.coronawarn.app/en/faq/#val_service_no_valid_dcc"</string>
+    <!-- XTXT: Certificate Selection Screen: Provider Requirements Title -->
+    <string name="dcc_ticketing_certificate_selection_provider_requirements_title">"Anforderungen des Anbieters:"</string>
+    <!-- XTXT: Certificate Selection Screen: Person Birthday -->
+    <string name="dcc_ticketing_certificate_birthday">"Geburtsdatum: %s"</string>
+    <!-- XTXT: Certificate Selection Screen: Certificates That Meet Requirements Title -->
+    <string name="dcc_ticketing_certificates_meeting_requirements_title">"Folgende Zertifikate entsprechen den Anforderungen des Anbieters und können zum Nachweis verwendet werden:"</string>
+    <!-- XTXT: Certificate Selection Screen: Rapid anitgen Test certificate-->
+    <string name="rat_test_certificate">Shcnelltest-Testzertifikat</string>
```
Comment (BMItr) [inline Corona-Warn-App/src/main/res/values-de/dcc_ticketing_strings.xml:83]: please correct to: "Schnelltest"

File: Corona-Warn-App/src/main/res/values/dcc_ticketing_strings.xml [hunk 18]
```diff
@@ -55,4 +55,32 @@
...
+    <!--XTXT: Certificate Selection Screen: FAQ Link EN -->
+    <string name="dcc_ticketing_faq_link_english">"https://www.coronawarn.app/en/faq/#val_service_no_valid_dcc"</string>
+    <!-- XTXT: Certificate Selection Screen: Provider Requirements Title -->
+    <string name="dcc_ticketing_certificate_selection_provider_requirements_title">"Anforderungen des Anbieters:"</string>
+    <!-- XTXT: Certificate Selection Screen: Person Birthday -->
+    <string name="dcc_ticketing_certificate_birthday">"Geburtsdatum: %s"</string>
+    <!-- XTXT: Certificate Selection Screen: Certificates That Meet Requirements Title -->
+    <string name="dcc_ticketing_certificates_meeting_requirements_title">"Folgende Zertifikate entsprechen den Anforderungen des Anbieters und können zum Nachweis verwendet werden:"</string>
+    <!-- XTXT: Certificate Selection Screen: Rapid anitgen Test certificate-->
+    <string name="rat_test_certificate">Shcnelltest-Testzertifikat</string>
```
Comment (BMItr) [inline Corona-Warn-App/src/main/res/values/dcc_ticketing_strings.xml:83]: please correct to: "Schnelltest"

**Review by AlexanderAlferov — DISMISSED** [review]
LGTM 🚀

File: Corona-Warn-App/src/main/res/layout/dcc_ticketing_valid_certificate_header_card.xml [hunk 19]
```diff
@@ -0,0 +1,62 @@
...
+    <TextView
+        android:id="@+id/allowed_certificates"
+        style="@style/body2Medium"
+        android:layout_width="0dp"
+        android:layout_height="wrap_content"
+        android:layout_marginTop="8dp"
+        app:layout_constraintEnd_toEndOf="@id/selection_title"
+        app:layout_constraintStart_toStartOf="@id/selection_title"
+        app:layout_constraintTop_toBottomOf="@id/selection_title"
+        tools:text="Impfzertifikat, Genesenenzertifikat, Shcnelltest-Testzertifikat, PCR-Testzertifikat" />
```
Comment (BMItr) [inline Corona-Warn-App/src/main/res/layout/dcc_ticketing_valid_certificate_header_card.xml:27]: please correct to: "Schnelltest"

**Review by BMItr — APPROVED** [review]

**Review by AlexanderAlferov — APPROVED** [review]
